lib/joinserver/joinserver_test.go

```diff
@@ -303,6 +321,123 @@ func TestJoinServiceGRPCServer_RegisterUsingAzureMethod(t *testing.T) {
+		initReq              *proto.RegisterUsingTPMMethodInitialRequest
+		challengeResponse    *proto.RegisterUsingTPMMethodChallengeResponse
+		challengeResponseErr error
+		authErr              error
+		certs                *proto.Certs
+	}{
+		{
+			desc: "pass case",
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("llama"),
+				},
+				JoinRequest: &types.RegisterUsingTokenRequest{
+					Token: "xyzzy",
+				},
+			},
+			challenge: "foo",
+			challengeResponse: &proto.RegisterUsingTPMMethodChallengeResponse{
+				Solution: []byte("bar"),
+			},
+			certs: &proto.Certs{SSH: []byte("qux")},
+		},
+		{
+			desc: "auth error",
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("llama"),
+				},
+				JoinRequest: &types.RegisterUsingTokenRequest{
+					Token: "xyzzy",
+				},
+			},
+			challenge: "foo",
+			challengeResponse: &proto.RegisterUsingTPMMethodChallengeResponse{
+				Solution: []byte("bar"),
+			},
+			authErr: trace.AccessDenied("test auth error"),
+		},
+		{
+			desc: "challenge response error",
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("llama"),
+				},
+				JoinRequest: &types.RegisterUsingTokenRequest{
+					Token: "xyzzy",
+				},
+			},
+			challenge:            "foo",
+			challengeResponseErr: trace.BadParameter("test challenge error"),
+		},
+		{
+			desc: "missing join request",
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("llama"),
+				},
+				JoinRequest: nil,
+			},
+			challenge: "foo",
+			authErr: trace.BadParameter(
+				"expected JoinRequest in RegisterUsingTPMMethodRequest_Init, got nil",
+			),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			testPack.mockAuthServer.sendChallenge = tc.challenge
+			testPack.mockAuthServer.returnCerts = tc.certs
+			testPack.mockAuthServer.returnError = tc.authErr
+			challengeResponder := func(
+				challenge *proto.TPMEncryptedCredential,
+			) (*proto.RegisterUsingTPMMethodChallengeResponse, error) {
+				require.Equal(t, &proto.TPMEncryptedCredential{
+					Secret: []byte(tc.challenge),
+				}, challenge)
+				return tc.challengeResponse, tc.challengeResponseErr
+			}
+
+			for suffix, clt := range map[string]*client.JoinServiceClient{
+				"_auth":  testPack.authClient,
+				"_proxy": testPack.proxyClient,
+			} {
+				t.Run(tc.desc+suffix, func(t *testing.T) {
+					certs, err := clt.RegisterUsingTPMMethod(
+						context.Background(), tc.initReq, challengeResponder,
+					)
+					if tc.challengeResponseErr != nil {
+						require.ErrorIs(t, err, tc.challengeResponseErr)
+						return
+					}
+					if tc.authErr != nil {
+						require.Contains(t, err.Error(), tc.authErr.Error())
```

Comment:
If we only care about the message, then make authErr a string and use require.ErrorContains?

```go
						require.ErrorContains(t, err.Error(), tc.authErr, "authErr mismatch")
```

Reply:
c3e2564dc5